plugin/k8s_external/external.go

```diff
@@ -68,8 +73,12 @@ func (e *External) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
 		return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
 	}
 
-	if e.externalFunc == nil {
-		return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
+	if !e.kubernetesPluginLoaded {
+		if e.Fall.Through(state.Name()) {
+			return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
+		}
```

Comment:
> I appreciate the added error here. If the kubernetes plugin is not loaded, the configuration is essentially invalid, the user needs to know about it and fix it, but I think we should not fall through.

I also think so, but previously, if the Kubernetes plugin was not loaded, we defaulted to doing a fall through for the user. If we don't do a fall through now, we should return this error in advance and prevent the program from starting during setup phase instead of just logging it?

Here, I am complying with the previous agreement where 'fall through' is the default for users. However, it is now necessary to enable the `fallthrough` option.

Reply:
> we should return this error in advance and prevent the program from starting during setup phase 

If we could that would be ideal, but IIRC we cannot detect this during setup due to the plugin initialization order. We have to wait until all plugins are loaded, so we defer this check to onStartup.

> Here, I am complying with the previous agreement where 'fall through' is the default for users.
> However, it is now necessary to enable the fallthrough option.

If the configuration is bad (missing kubernetes), the default is to fall through to next plugin.  _I think this existing behavior is flawed_, it hides a broken configuration from the user, and would never result in a running-as-intended instance of CoreDNS. It is better to error when we know the configuration is wrong than to be silent and behave in a way that we know is not intended - e.g. sending queries upstream that could result in wrong answers sent to a client.

I appreciate and prefer your addition of adding an error here.  IMO, you are fixing a bug.  But again, I think the error should always be returned when the configuration is bad, even when the user is using `fallthrough`.

Reply:
Thank you for the explanation. I've also observed that enabling the `errors` plugin results in an extra error log being printed twice. Furthermore, I've made changes to the default behavior such that if Kubernetes isn't loaded, we'll return a `SERVFAIL` response instead of proceeding to the next plugin.

After reviewing the code and testing, I suggest we check for this issue during startup and failed start attempts. I will address this in a future pull request. Does that sound good to you?

Reply:
> After reviewing the code and testing, I suggest we check for this issue during startup and failed start attempts. I will address this in a future pull request. Does that sound good to you?

This isn't possible for the reason I explained above.